rust/src/index/vector/diskann/builder.rs

```diff
@@ -0,0 +1,494 @@
+            continue;
+        }
+        visited.insert(p.id);
+        for neighbor_id in graph.neighbors(p.id).iter() {
+            let neighbor_id = *neighbor_id as usize;
+            if visited.contains(&neighbor_id) {
+                // Already visited.
+                continue;
+            }
+            let dist = distance_to(vectors, query, neighbor_id)?;
+            candidates.insert(OrderedFloat(dist), neighbor_id as usize);
+            if candidates.len() > search_size {
+                candidates.pop_last();
+            }
+        }
+    }
+
+    Ok((
+        candidates.iter().take(k).map(|(_, id)| *id).collect(),
+        visited,
+    ))
+}
+
+/// Algorithm 2 in the paper.
+async fn robust_prune(
+    graph: &GraphBuilder<PQVertex, DiskANNParams>,
+    matrix: &MatrixView,
+    id: usize,
+    mut visited: HashSet<usize>,
+    alpha: f32,
+    r: usize,
+) -> Result<Vec<u32>> {
+    visited.remove(&id);
+    let neighbors = graph.neighbors(id);
+    visited.extend(neighbors.iter().map(|id| *id as usize));
+
+    let mut heap: BinaryHeap<VertexWithDistance> = visited
+        .iter()
+        .map(|v| {
+            let dist = distance(matrix, id, *v).unwrap();
+            VertexWithDistance {
+                id: *v,
+                distance: OrderedFloat(dist),
+            }
+        })
+        .collect();
+
+    let matrix = matrix.clone();
+    let new_neighbours = tokio::task::spawn_blocking(move || {
+        let mut new_neighbours: Vec<usize> = vec![];
+        while !visited.is_empty() {
+            let mut p = heap.pop().unwrap();
+            while !visited.contains(&p.id) {
+                // Because we are using a heap for `argmin(Visited)` in the original
+                // algorithm, we need to pop out the vertices that are not in `visited` anymore.
+                p = heap.pop().unwrap();
+            }
+
+            new_neighbours.push(p.id);
+            if new_neighbours.len() >= r {
+                break;
+            }
+
+            let mut to_remove: HashSet<usize> = HashSet::new();
+            for pv in visited.iter() {
+                let dist_prime = distance(&matrix, p.id, *pv)?;
+                let dist_query = distance(&matrix, id, *pv)?;
+                if alpha * dist_prime <= dist_query {
+                    to_remove.insert(*pv);
+                }
+            }
+            for pv in to_remove.iter() {
+                visited.remove(pv);
+            }
+        }
+        Ok::<_, Error>(new_neighbours)
+    })
+    .await??;
+
+    Ok(new_neighbours.iter().map(|id| *id as u32).collect())
+}
+
+/// Find the index of the medoid vector in all vectors.
+async fn find_medoid(vectors: &MatrixView, metric_type: MetricType) -> Result<usize> {
+    let centroid = vectors
+        .centroid()
+        .ok_or_else(|| Error::Index("Cannot find the medoid of an empty matrix".to_string()))?;
+
+    let dist_func = metric_type.func();
+    // Find the closest vertex to the centroid.
```

Comment:
They seems to start to use "closest one to centroid" since the `FreshDiskANN` paper, maybe due to the time complexity of finding medoid.

I can change the function name tho.